00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,373 @@
+```
+
+# C++ Wrapper API
+
+The current view is that the API exported by the Node.js binary
+must be in C in order to achieve ABI stability. This [document]
+(http://www.oracle.com/technetwork/articles/servers-storage-dev/stablecplusplusabi-333927.html)
+outlines some of the challenges with respect to C++ and
+ABI stability.
+
+However, we also believe that a C++ wrapper is possible and
+would provide ease of use. The key element is that the C++ part must all
+be inlined and only use the external functions/types exported
+by the C API.  
+
+What follows are some examples (incomplete/immature) mirroring NaN when
+possible on what some of the C++ wrapper might include. Its not as far along
+as the C API as we've only added things that were needed in porting the
+modules we've worked on so far.  In the end we expect a more complete set.
+
+```C++
+  class HandleScope {
+  public:
+    HandleScope() {
+    explicit HandleScope(napi_env e) 
+    ~HandleScope()
+  };
+
+  class EscapableHandleScope {
+  public:
+    EscapableHandleScope() 
+    explicit EscapableHandleScope(napi_env e)
+    ~EscapableHandleScope() 
+  
+    napi_value Escape
+  };
+
+  class Utf8String {
+  public:
+    inline explicit Utf8String(napi_value from) 
+    inline int length() 
+    inline char* operator*() { return str_; }
+    inline const char* operator*() const { return str_; }
+
+    inline ~Utf8String()
+  };
+
+  class Callback {
+   public:
+    Callback() 
+    explicit Callback(napi_value fn)
+    ~Callback() 
+
+    bool operator==(const Callback &other) 
+    bool operator!=(const Callback &other) 
+    inline napi_value operator*() 
+    inline napi_value operator()
+    inline void SetFunction(napi_value fn) 
+    inline napi_value GetFunction() 
+    inline bool IsEmpty() 
+    inline napi_value
+    Call(napi_value target,
+        int argc,
+        napi_value argv[]) 
+    inline napi_value
+    Call(int argc, napi_value argv[]) 
+  };
+
+  /* abstract */ class AsyncWorker {
+   public:
+    explicit AsyncWorker(Callback *callback_)
+    virtual ~AsyncWorker()
+    virtual void WorkComplete() 
+    inline void SaveToPersistent(
+        const char *key, napi_value value)
+    inline void SaveToPersistent(
+        napi_propertyname key, napi_value value) 
+    inline void SaveToPersistent(
+        uint32_t index, napi_value value)
+    inline napi_value GetFromPersistent(const char *key) 
+
+    inline napi_value
+    GetFromPersistent(napi_propertyname key) 
+
+    inline napi_value GetFromPersistent(uint32_t index) 
+    virtual void Execute() = 0;
+    virtual void Destroy() 
+    
+  protected:
+    virtual void HandleOKCallback() 
+    virtual void HandleErrorCallback()
+    void SetErrorMessage(const char *msg) 
+
+    const char* ErrorMessage() 
+  };
+
+  inline void AsyncExecute(uv_work_t* req)
+  inline void AsyncExecuteComplete(uv_work_t* req)
+  inline void AsyncQueueWorker(AsyncWorker* worker) 
+```
```

Comment:
I am a bit conflicted about `AsyncWorker`. We are currently making it "part of core", when it is currently not.
However it is highly used in C++ addons, so I think it will be highly useful.

Can we keep all of these in another module for the time being? It's not bad to provide those via a dependency that we can version independently.